latch/account.py

```diff
@@ -1,45 +1,95 @@
 from latch.registry.project import Project
+from latch_cli.config.user import user_config
+
+
+class _CatalogProjectNode(TypedDict):
+    id: str
+    displayName: str
+
+
+@dataclass
+class _Cache:
+    catalog_projects: Optional[List[Project]] = None
 
 
 @dataclass(frozen=True)
 class Account:
+    _cache: _Cache = field(
+        default_factory=lambda: _Cache(),
+        init=False,
+        repr=False,
+        hash=False,
+        compare=False,
+    )
+
     id: str
 
     @classmethod
+    @cache
     def current(cls):
-        account_id = execute(
-            document="""
-                query accountInfoQuery {
-                    accountInfoCurrent {
-                        id
+        if user_config.workspace != "":
+            account_id = user_config.workspace
+        else:
+            account_id = execute(
+                document=gql.gql("""
+                    query accountInfoQuery {
+                        accountInfoCurrent {
+                            id
+                        }
                     }
-                }
-            """,
-        )["accountInfoCurrent"]["id"]
+                    """),
+            )["accountInfoCurrent"]["id"]
 
         return cls(id=account_id)
 
-    def list_projects(self) -> List[Project]:
-        query = f"""
-            query ProjectsQuery ($argOwnerId: BigInt!) {{
-                catalogProjects (
-                    condition: {{
-                        ownerId: $argOwnerId
-                        removed: false
-                    }}
-                ) {{
-                    nodes {{
-                        id
-                        displayName
-                    }}
-                }}
-            }}
-        """
-
-        data = execute(query, {"argOwnerId": self.id})
-
-        return [Project(node["id"]) for node in data["catalogProjects"]["nodes"]]
+    def load(self):
+        data = execute(
+            gql.gql("""
+            query AccountQuery($ownerId: BigInt!) {
+                accountInfo(id: $ownerId) {
+                    catalogProjectsByOwnerId(
+                        condition: {
+                            removed: false
+                        }
+                    ) {
+                        nodes {
+                            id
+                            displayName
+                        }
+                    }
+                }
+            }
+            """),
+            {"ownerId": self.id},
+        )["accountInfo"]
+        # todo(maximsmol): deal with nonexistent accounts
+
+        self._cache.catalog_projects = []
+        x: _CatalogProjectNode
+        for x in data["catalogProjectsByOwnerId"]["nodes"]:
+            cur = Project(x["id"])
+            cur._cache.display_name = x["displayName"]
+            self._cache.catalog_projects.append(cur)
+
+    # list_projects
+
+    @overload
+    def list_projects(self, *, load_if_missing: Literal[True] = True) -> List[Project]:
+        ...
```

Comment:
So `None` indicates and unloaded state and an empty array [] indicates no projects? Not sure why `load_if_missing` is supported at all.

Reply:
It would make more sense to have a "network-only" mode or something which forces requests.

Reply:
how does that not make sense? seems fairly natural for `None` to mean something hasn't loaded, and `[]` to mean that that something has loaded, and that it is empty

`network-only` would only make sense if there was a pressing need to refresh values, i.e. if the cache became stale fairly regularly. at the granularity of Projects/Tables, we're making the assumption that such things only change sparsely

Reply:
you can also refetch by calling `.load()` again or by making a new `Table`

Reply:
but why would anyone call it with `load_if_missing` set to false?

Reply:
per max: 'standard API design principles: if you have a method that requires a blocking operation, have a non-blocking version of it'

Reply:
also see the `__str__` methods for use cases for `load_if_missing=False`